### PR TITLE
Pay using Apple Pay

### PR DESCRIPTION
**Pull request by TheGrimmChester**

### CODE REVIEW

File: src/Resources/views/SyliusShopBundle/Checkout/SelectPayment/_choice.html.twig [hunk 1]
```diff
@@ -62,17 +65,141 @@
...
             </div>
         {% endif %}
+
+        {% if method.gatewayConfig.factoryName == applePayFactoryName %}
+            <script src="https://applepay.cdn-apple.com/jsapi/v1/apple-pay-sdk.js"></script>
+
+            <apple-pay-button id="payplug-applepay-btn" onclick="apple();" buttonstyle="black" type="plain" locale="en"></apple-pay-button>
+            <div>
+                <script>
+                    function apple() {
```
Comment (TheGrimmChester) [inline src/Resources/views/SyliusShopBundle/Checkout/SelectPayment/_choice.html.twig:87]: TODO: Front refacto

**Review by tuyetrinhvo — APPROVED** [review]

File: src/Controller/OrderController.php [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+            if (PaymentInterface::STATE_COMPLETED !== $lastPayment->getState()) {
+                throw new PaymentNotCompletedException();
+            }
+        } catch (\Exception|PaymentNotCompletedException $exception) {
+            $this->addFlash('error', 'sylius.payment.cancelled');
+            $redirect = $this->redirectToRoute('sylius_shop_checkout_select_payment');
+            $dataResponse = [];
+            $dataResponse['returnUrl'] = $redirect->getTargetUrl();
+            $dataResponse['responseToApple'] = ['status' => 0];
```
Comment (oallain) [inline src/Controller/OrderController.php:191]: Can you use const for `0` too ?

File: src/Controller/OrderController.php [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        }
+
+        $order = $lastPayment->getOrder();
+        Assert::isInstanceOf($order, OrderInterface::class);
+
+        $request->getSession()->set('sylius_order_id', $order->getId());
+        $dataResponse = [];
+        $redirect = $this->redirectToRoute('sylius_shop_checkout_complete');
+        $dataResponse['returnUrl'] = $redirect->getTargetUrl();
+        $dataResponse['responseToApple'] = ['status' => 1];
```
Comment (oallain) [inline src/Controller/OrderController.php:225]: Same here.

File: src/Controller/OrderController.php [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+            $currentCart = $this->getCurrentCart();
+
+            $this->applePayPaymentProvider->cancel($currentCart);
+
+            $this->addFlash('error', 'sylius.payment.cancelled');
+
+            $dataResponse = [];
+            $redirect = $this->redirectToRoute('sylius_shop_checkout_select_payment');
+            $dataResponse['returnUrl'] = $redirect->getTargetUrl();
+            $dataResponse['responseToApple'] = ['status' => 1];
```
Comment (oallain) [inline src/Controller/OrderController.php:287]: Same here

File: src/Controller/OrderController.php [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        } catch (\Exception $exception) {
+            $this->addFlash('error', 'sylius.payment.cancelled');
+
+            $dataResponse = [];
+            $redirect = $this->redirectToRoute('sylius_shop_checkout_select_payment');
+            $dataResponse['returnUrl'] = $redirect->getTargetUrl();
+            $dataResponse['responseToApple'] = ['status' => 1];
+
+            $response = [
+                'success' => true,
```
Comment (oallain) [inline src/Controller/OrderController.php:306]: Same here

**Review by oallain — APPROVED** [review]

**Pull request merged**